src/establish-metamask-repository.ts

```diff
@@ -0,0 +1,259 @@
+import {
+  directoryExists,
+  ensureDirectoryStructureExists,
+} from '@metamask/utils/node';
+import execa from 'execa';
+import fs from 'fs';
+import path from 'path';
+
+import { createModuleLogger, projectLogger } from './logging-utils';
+import type { AbstractOutputLogger } from './output-logger';
+import { RepositoryFilesystem } from './repository-filesystem';
+import {
+  ensureDefaultBranchIsUpToDate,
+  getBranchInfo,
+  getCurrentBranchName,
+} from './repository-utils';
+import type { BranchInfo } from './repository-utils';
+import { resolveRepositoryReference } from './resolve-repository-reference';
+
+const log = createModuleLogger(projectLogger, 'establish-metamask-repository');
+
+type ConfirmedRepository = BranchInfo & {
+  shortname: string;
+  directoryPath: string;
+};
+
+/**
+ * A repository within the MetaMask organization on GitHub which has been cloned
+ * to the local filesystem. More concretely, this could either be a template or
+ * a project that requires linting.
+ */
+export type MetaMaskRepository = ConfirmedRepository & {
+  fs: RepositoryFilesystem;
+};
+
+/**
+ * Information about a repository we know exists on the filesystem.
+ */
+export type ExistingRepository = ConfirmedRepository & {
+  isKnownMetaMaskRepository: boolean;
+};
+
+/**
+ * Ensures that there is a proper repository to lint. A repository may be one of
+ * two things: either A) the "short name" of a known repository under the GitHub
```

Comment:
What is the purpose of these two modes? e.g. when would a short name be passed in, and when would a directory path be provided

Reply:
This would depend on whether this tool is being used outside of any project or inside a specific project.

As it stands, I expect this tool to be primarily used outside of a particular project. This would be the easiest way to use it. For instance, I want to lint `utils`, so I just pass `utils` on the command line, and the tool clones the repository automatically; I don't need to clone it myself. Or I pass in nothing, in which case the tool clones all of the known repositories. Either way, this function would get passed a short name.

However, say you had `utils` cloned already, and you were working on addressing lint violations, and you wanted to re-run `module-lint` against your changes to see how it changed the report. In that case you could just pass your clone directory to `module-lint`, and it would use this instead of using _its_ clone of `utils`.

This latter use case also opens the door to a `--fix` option which, similar to ESLint, would attempt to autofix lint violations. In that case you'd need to run this tool in the context of a local directory and not a temporary clone, otherwise those fixes wouldn't be very accessible.

Reply:
In the latter case, the "ensureDefaultBranchIsUpToDate" step seems undesirable. You wouldn't want a linter to throw away your changes or introduce merge conflicts

Reply:
Absolutely! There is a check in `establishMetaMaskRepository` which should ensure that this step only happens if we have a known MetaMask repository that we know has already been cloned or will be cloned: https://github.com/MetaMask/module-lint/pull/30/files#diff-31cf46dfa3c2406e1ddba3021f532dc0c0cbebfdcbf4726a567e08d4fa99329fR84. That said, I realize looking back at this that there is a bug with this, so I will push up a fix.

Reply:
Ah. That's fair. I've renamed it to `createdAutomatically` here: [0afaf74](https://github.com/MetaMask/module-lint/pull/30/commits/0afaf742b406e0ece94436688653122c26daf3a7). Does that help?

Reply:
Thanks that is much better

Reply:
Nit: Thoughts on making that same change elsewhere as well? It seems like `createdAutomatically` does describe the situation better even in the resolve step.

And it gets rid of that "and assumes that the directory is not a clone of a known MetaMask repository" part of the `resolveRepositoryReference` test. When reading that I wondered why we were assuming that, given that we _would_ expect to use this tool on local MetaMask repositories.

Reply:
Makes sense. Updated here: [f7b684d](https://github.com/MetaMask/module-lint/pull/30/commits/f7b684d7563d6affc11bec4d2a63aade761e4a44)